gcd/src/main/java/org/spine3/server/storage/datastore/DsCommandStorage.java

```diff
@@ -37,31 +44,95 @@
+        checkNotNull(commandId);
+
+        final CommandStorageRecord updatedRecord = read(commandId)
+                .toBuilder()
+                .setStatus(CommandStatus.OK)
+                .build();
+        write(commandId, updatedRecord);
+    }
+
+    @Override
+    public void updateStatus(CommandId commandId, Error error) {
+        checkNotNull(commandId);
+        checkNotNull(error);
+
+        final CommandStorageRecord updatedRecord = read(commandId)
+                .toBuilder()
+                .setStatus(CommandStatus.ERROR)
+                .setError(error)
+                .build();
+        write(commandId, updatedRecord);
+    }
+
+    @Override
+    public void updateStatus(CommandId commandId, Failure failure) {
+        checkNotNull(commandId);
+        checkNotNull(failure);
+
+        final CommandStorageRecord updatedRecord = read(commandId)
+                .toBuilder()
+                .setStatus(CommandStatus.FAILURE)
+                .setFailure(failure)
+                .build();
+        write(commandId, updatedRecord);
     }
 
     @Override
-    protected void write(CommandStoreRecord record) {
+    public CommandStorageRecord read(CommandId commandId) {
+        checkNotClosed();
+        checkNotDefault(commandId);
+
+        final String idString = idToString(commandId);
+        final Key.Builder key = createKey(idString);
+        final LookupRequest request = LookupRequest.newBuilder().addKey(key).build();
+
+        final LookupResponse response = datastore.lookup(request);
 
-        final Value.Builder id = makeValue(record.getCommandId());
-        final Property.Builder idProperty = makeProperty(COMMAND_ID_PROPERTY_NAME, id);
-        final Entity.Builder entity = messageToEntity(record, makeKey(KIND));
-        entity.addProperty(idProperty);
-        entity.addProperty(makeTimestampProperty(record.getTimestamp()));
+        if (response == null || response.getFoundCount() == 0) {
+            return CommandStorageRecord.getDefaultInstance();
+        }
 
-        final Mutation.Builder mutation = Mutation.newBuilder().addInsertAutoId(entity);
+        final EntityResult entity = response.getFound(0);
+        final CommandStorageRecord result = entityToMessage(entity, typeName.toTypeUrl());
+        return result;
+    }
+
+    @Override
+    public void write(CommandId commandId, CommandStorageRecord record) {
+        checkNotClosed();
+        checkNotDefault(commandId);
+        checkNotDefault(record);
+
+        final String idString = idToString(commandId);
+
+        final Key.Builder key = createKey(idString);
+
+        final Entity.Builder entity = messageToEntity(record, key);
+        entity.addProperty(DatastoreProperties.makeTimestampProperty(record.getTimestamp()));
+        entity.addProperty(DatastoreProperties.makeTimestampNanosProperty(record.getTimestamp()));
+
+        final Mutation.Builder mutation = Mutation.newBuilder().addUpsert(entity);
```

Comment:
INSERT + UPDATE = UPSERT
https://cloud.google.com/datastore/reference/rpc/google.datastore.v1beta3

Welcome to Google API naming conventions